### ISSUE TITLE
[BUG] pip install DeepSeed Error

### ISSUE THREAD

**Issue by ElbekJK** [microsoft/DeepSpeed#3145]
When I try to download the DeepSpeed with pip, I am constantly facing with this error. I would appreciate your help!

Collecting deepspeed
  Using cached deepspeed-0.8.3.tar.gz (765 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      test.c
      LINK : fatal error LNK1181: cannot open input file 'aio.lib'
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\Elbek\AppData\Local\Temp\pip-install-yduslt_a\deepspeed_eacd64e4be3945869afa52e220c455fa\setup.py", line 156, in <module>
          abort(f"Unable to pre-compile {op_name}")
        File "C:\Users\Elbek\AppData\Local\Temp\pip-install-yduslt_a\deepspeed_eacd64e4be3945869afa52e220c455fa\setup.py", line 48, in abort
          assert False, msg
      AssertionError: Unable to pre-compile async_io
      DS_BUILD_OPS=1
       [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
       [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
       [WARNING]  One can disable async_io with DS_BUILD_AIO=0
       [ERROR]  Unable to pre-compile async_io
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Comment by ElbekJK** [comment]
I am using Windows. But I cannot create a wheel to request `ds_report`.

**Comment by ElbekJK** [comment]
I was building ops. I did what you suggested but it didn't work either. `set DS_BUILD_OPS=0
PS C:\Users\Elbek\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\site-packages\deepspeed-0.8.3> python setup.py bdist_wheel
DS_BUILD_OPS=1
test.c
LINK : fatal error LNK1181: cannot open input file 'aio.lib'
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
 [WARNING]  One can disable async_io with DS_BUILD_AIO=0
 [ERROR]  Unable to pre-compile async_io`

**Comment by ElbekJK** [comment]
Yeah, it worked. Thanks a lot. But I don't know what to do from here. Because I don't know if the deepspeed is working.

**Comment by GalaxyHe2023** [comment]
I reinstalled the OS, re-downloaded cuda and cudann, made sure my cuda and cudann are both certified.

set DS_BUILD_AIO=0

set DS_BUILD_SPARSE_ATTN=0

Then the error evolved.

No CUDA runtime is found, using CUDA_HOME='E:\00-chatGPT\CUDA'
The system can not find the file specified.
The system can not find the file specified.
The system can not find the file specified.
The system can not find the file specified.
fatal: not a git repository (or any of the parent directories): .git
Traceback (most recent call last):
File "", line 2, in
File "", line 34, in
File "C:\Users\Administrator\AppData\Local\Temp\pip-install-nqtc3o4j\deepspeed_d3894dbfb2fe4c1984b1cd4a0b06e0f7\setup.py", line 198, in
create_dir_symlink('..\..\csrc', '.\deepspeed\ops\csrc')
File "C:\Users\Administrator\AppData\Local\Temp\pip-install-nqtc3o4j\deepspeed_d3894dbfb2fe4c1984b1cd4a0b06e0f7\setup.py", line 190, in create_dir_symlink
os. remove(dest)
PermissionError: [WinError 5] Access is denied. : '.\deepspeed\ops\csrc'

Are there other solutions?

**Comment by GalaxyHe2023** [comment]
I use wsl to fix the issue.

**Comment by lsm03624** [comment]
> 啊，如果您使用的是PowerShell，则需要以不同的方式设置环境变量，我假设这是cmd.exe。对于PowerShell，您可以尝试吗`Set-Item Env:\DS_BUILD_OPS 0`

After setting it up this way, the compilation didn't take place and the directly generated .whl file couldn't be installed.

**Comment by lsm03624** [comment]
> Ah, if you are using PowerShell you'll want to set the environment variable differently, I assumed that was cmd.exe. For PowerShell can you try `Set-Item Env:\DS_BUILD_OPS 0`

After setting it up this way, the compilation didn't take place and the directly generated .whl file couldn't be installed.

**Comment by ccu1tn** [comment]
@loadams Hello. I also have the same problem. here is my problem. I don't know how to fix it. 
I built on window with vscode.
DeepSpeed: git clone from https://github.com/microsoft/DeepSpeed
(speech_env_one) is a virtual environment.

(speech_env_one) D:\Attack\SemCom\SPEECH_TO_TEXT\DeepSC-ST_demonstration\DeepSpeed>python setup.py bdist_wheel
DS_BUILD_OPS=1
test.c
LINK : fatal error LNK1181: cannot open input file 'aio.lib'
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
 [WARNING]  One can disable async_io with DS_BUILD_AIO=0
 [ERROR]  Unable to pre-compile async_io
Traceback (most recent call last):
  File "setup.py", line 182, in <module>
    abort(f"Unable to pre-compile {op_name}")
  File "setup.py", line 52, in abort
    assert False, msg
AssertionError: Unable to pre-compile async_io

(speech_env_one) D:\Attack\SemCom\SPEECH_TO_TEXT\DeepSC-ST_demonstration\DeepSpeed>Set-Item Env:\DS_BUILD_OPS 0
'Set-Item' is not recognized as an internal or external command,
operable program or batch file.

**Comment by loadams** [comment]
Hi @ccu1tn - the failure to set the environment variable seems that you are not in a Powershell environment since the powershell commands are not working currently for you?  Can you confirm if you're able to open a new powershell window and run that?

**Comment by ccu1tn** [comment]
Hi @loadams, thank you for your support. It worked when I ran with PowerShell. The results are like this, Is it correct? 
![image](https://github.com/microsoft/DeepSpeed/assets/43333942/3676adf5-bece-42cc-9c01-2ee3106c356b)

I also have another error in this picture. 
TensorFlow: 2.3.1/
Cuda 10.1.
python 3.7. 
I fixed it by re-installing TensorFlow 2.2. It still does not work. 
![image](https://github.com/microsoft/DeepSpeed/assets/43333942/28aeeb51-ff0f-45ce-9490-796cad9c33c8)

**Comment by melMass** [comment]
I just went down the rabbit hole of #4729 #4669 and I'm stuck on this issue specified here and hesitant to open a new one:

![image](https://github.com/microsoft/DeepSpeed/assets/7041726/b04df922-5c23-47ea-b4e1-b9e6bc479c2f)

<details><summary>as text for SEO</summary>

```sh
❯ pip install deepspeed --pre
Collecting deepspeed
  Using cached deepspeed-0.12.6.tar.gz (1.2 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [36 lines of output]
      test.c
      LINK : fatal error LNK1181: cannot open input file 'aio.lib'
      The system cannot find the file specified.
      The system cannot find the file specified.
      The system cannot find the file specified.
      The system cannot find the file specified.
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\User\AppData\Local\Temp\pip-install-zl23ifx7\deepspeed_74d1c763267644b39987945a8d63c94f\setup.py", line 224, in <module>
          create_dir_symlink('..\\..\\csrc', '.\\deepspeed\\ops\\csrc')
        File "C:\Users\User\AppData\Local\Temp\pip-install-zl23ifx7\deepspeed_74d1c763267644b39987945a8d63c94f\setup.py", line 216, in create_dir_symlink
          os.remove(dest)
      PermissionError: [WinError 5] Access is denied: '.\\deepspeed\\ops\\csrc'
      DS_BUILD_OPS=1
       [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
       [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
       [WARNING]  cpu_adam requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adam attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adam requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adam attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adagrad requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adagrad attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_adagrad requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_adagrad attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_lion requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_lion attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  cpu_lion requires the 'lscpu' command, but it does not exist!
       [WARNING]  cpu_lion attempted to query 'lscpu' after failing to use py-cpuinfo to detect the CPU architecture. 'lscpu' does not appear to exist on your system, will fall back to use -march=native and non-vectorized execution.
       [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
       [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
       [WARNING]  please install triton==1.0.0 if you want to use sparse attention
      Install Ops={'async_io': False, 'fused_adam': 1, 'cpu_adam': 1, 'cpu_adagrad': 1, 'cpu_lion': 1, 'evoformer_attn': False, 'fused_lamb': 1, 'fused_lion': 1, 'inference_core_ops': 1, 'cutlass_ops': False, 'quantizer': 1, 'ragged_device_ops': False, 'ragged_ops': 1, 'random_ltd': 1, 'sparse_attn': False, 'spatial_inference': 1, 'transformer': 1, 'stochastic_transformer': 1, 'transformer_inference': 1}
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

</details>

**Comment by melMass** [comment]
It works if I clone the repo, I published wheels here for windows

https://github.com/melMass/DeepSpeed/releases/tag/v0.12.7

**Comment by szriru** [comment]
@melMass 
Thanks :)

**Comment by XiangPiIi** [comment]
> It works if I clone the repo, I published wheels here for windows
> 
> https://github.com/melMass/DeepSpeed/releases/tag/v0.12.7

nice work！

**Comment by nicotie** [comment]
> It works if I clone the repo, I published wheels here for windows
> 
> https://github.com/melMass/DeepSpeed/releases/tag/v0.12.7

Thanks!

**Comment by teanhow** [comment]
> It works if I clone the repo, I published wheels here for windows
> 
> https://github.com/melMass/DeepSpeed/releases/tag/v0.12.7

Thanks :)

**Comment by xieliaing** [comment]
> It works if I clone the repo, I published wheels here for windows
> 
> https://github.com/melMass/DeepSpeed/releases/tag/v0.12.7

Thanks !

**Comment by niebowen666** [comment]
I have installed in Ubuntu 22.04，anaconda。but it has some errors：
[WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
[WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.

but I am sure that I have installed the libaio and libaio-dev：
libaio-dev is already the newest version (0.3.112-5).
libaio1 is already the newest version (0.3.112-5).
0 upgraded, 0 newly installed, 0 to remove and 23 not upgraded.
